Src/PerformanceCollector/Shared/QuickPulseTelemetryModule.cs

```diff
@@ -0,0 +1,468 @@
+                }
+            }
+        }
+
+        private void CollectionTimerCallback(object state)
+        {
+            try
+            {
+                this.CollectData();
+            }
+            catch (Exception e)
+            {
+                QuickPulseEventSource.Log.UnknownErrorEvent(e.ToString());
+            }
+            finally
+            {
+                // this is in a race condition with stopping timer from OnStopCollection, so we need to ensure that we don't schedule the next tick more than once
+                // after the timer has been ordered to stop
+                if (this.stateManager.IsCollectingData && this.collectionTimer != null)
+                {
+                    this.ScheduleNextCollection();
+                }
+            }
+        }
+        
+        private void CollectData()
+        {
+            var sample = this.CollectSample();
+
+            this.StoreSample(sample);
+        }
+
+        private void StoreSample(QuickPulseDataSample sample)
+        {
+            lock (this.collectedSamplesLock)
+            {
+                this.collectedSamples.AddLast(sample);
+
+                while (this.collectedSamples.Count > MaxSampleStorageSize)
+                {
+                    this.collectedSamples.RemoveFirst();
+                }
+            }
+        }
+
+        private QuickPulseDataSample CollectSample()
+        {
+            // For AI data, all we have to do is lock the current accumulator in
+            QuickPulseDataAccumulator completeAccumulator = this.dataAccumulatorManager.CompleteCurrentDataAccumulator();
+
+            // For performance collection, we have to read perf samples from Windows
+            List<Tuple<PerformanceCounterData, float>> perfData =
+                this.performanceCollector.Collect((counterName, e) => QuickPulseEventSource.Log.CounterReadingFailedEvent(e.ToString(), counterName))
+                    .ToList();
+
+            return this.CreateDataSample(completeAccumulator, perfData);
+        }
+
+        private QuickPulseDataSample CreateDataSample(
+            QuickPulseDataAccumulator accumulator,
+            IEnumerable<Tuple<PerformanceCounterData, float>> perfData)
+        {
+            return new QuickPulseDataSample(accumulator, perfData.ToDictionary(tuple => tuple.Item1.ReportAs, tuple => tuple));
+        }
+
+        #region Callbacks from the state manager
+        private void OnStartCollection()
+        {
+            this.dataAccumulatorManager.CompleteCurrentDataAccumulator();
+            this.telemetryProcessor.StartCollection(this.dataAccumulatorManager);
+
+            this.ScheduleNextCollection();
+        }
+
+        private void ScheduleNextCollection()
+        {
+            DateTime nextTick = this.collectionTimeSlotManager.GetNextCollectionTimeSlot(this.timeProvider.UtcNow);
+            this.collectionTimer.ScheduleNextTick((nextTick - this.timeProvider.UtcNow).Duration());
```

Comment:
this may throw if `GetNextCollectionTimeSlot` was called at 499 milliseconds and this line will be executed at 500 milliseconds.